src/authorize.ts

```diff
@@ -0,0 +1,175 @@
+import * as express from 'express';
+
+import * as session from './session';
+
+import debugModule from 'debug';
+const debug = debugModule('everyauth:authorize');
+const dbg = ({ userId, tenantId }: { userId?: string; tenantId?: string }, msg: string) => {
+  debug(`${tenantId || ''}${tenantId ? '/' : ''}${userId}: ${msg}`);
+};
+
+export interface IEveryAuthContext {
+  finishedUrl: string /** The url which the service was configured to use. */;
+
+  /**
+   * The identity that the user has been authenticated to; this is a database key internal to EveryAuth, and
+   * can be saved to reduce a lookup roundtrip on subsequent requests.
+   */
+  identityId: string;
+
+  tenantId: string /** The tenantId supplied by the mapToTenantId function for this request. */;
+  userId: string /** The userId supplied by the mapToUserId function for this request. */;
+}
+
+export interface IEveryAuthOptions {
+  /**
+   * The URL that the user should be sent to after the operation is complete. Can be either
+   * '/path/component/only' or a fully qualified URL.
+   *
+   * Supplied with an `error` query parameter if the operation resulted in an error.
+   */
+  finishedUrl: string;
+
+  /**
+   * Either a string or a function which generates a string that contains the protocol, hostname, port, and
+   * path to the current mount point.  If this is not supplied, the EveryAuth system attempts to determine it
+   * programmatically by looking at various attributes on the request object.
+   */
+  hostedBaseUrl?: string | ((req: express.Request) => string);
+
+  /**
+   * Map the current request to a unique tenantId that can be used to retrieve the authorized token on
+   * subsequent requests.  Usually extracts the value from the output of the authorization system, or can be a
+   * constant for experimentation and testing purposes.
+   *
+   * The tenant represents the larger customer, which may include multiple authorized users.
+   */
+  mapToTenantId?: ((req: express.Request) => Promise<string>) | ((req: express.Request) => string);
+
+  /**
+   * Map the current request to a unique userId tag that can be used to retrieve the authorized token on
+   * subsequent requests.  Usually extracts the value from the output of the authorization system, or can be a
+   * constant for experimentation and testing purposes.
+   */
+  mapToUserId: ((req: express.Request) => Promise<string>) | ((req: express.Request) => string);
```

Comment:
We talked today about how EveryAuth's value prop is very thin without any foreign keys. Having said that, I don't think we should *force* people to specify any tags, userId included. There is a corresponding comments in session.ts.

Reply:
Talked on Slack - let's not block the MVP on this, we can make it optional later if needed.